test/differentials/composite.jl

```diff
@@ -200,6 +200,41 @@ end
             d_sum = Composite{Dict}(Dict(4 => 3.0 + 3.0, 3 => 2.0, 2 => 2.0))
             @test d1 + d2 == d_sum
         end
+
+        @testset "Fields of type NotImplemented" begin
+            CFoo = Composite{Foo}
+            a = CFoo(x=1.5)
+            b = CFoo(x=@not_implemented(""))
+            for (x, y) in ((a, b), (b, a), (b, b))
```

Comment:
We probably don't need to test exhaustively? IMO we could just test `Composite{Foo}` with `(a, b), (b, a)` here.

Reply:
`(b, b)` tests `+(::NotImplemented, ::NotImplemented)` but, as the other two cases, in principle this is covered by the `NotImplemented` tests now as well. I kept the tests of the original PR since this was the actual problem that caused errors in ChainRulesTestUtils. Would you like me to remove the `Composite` tests?

Reply:
That would be my preference, but it is not strong and I leave it up to you to make a decision. Feel free to merge as is!

Reply:
If it's not necessary to remove the tests, I would like to keep them since they check the actual problem that this PR tries to solve. I assume that this is an advantage if, e.g., at some point a more restrictive policy regarding summation of `NotImplemented` is adopted.